### DISCUSSION TITLE
Remote sensor fails to authenticate, manager mine.p does not exist

### DISCUSSION THREAD

**Discussion by darth-drew** [Security-Onion-Solutions/securityonion#7358] [Unsupported Versions]
Currently running a 2.3.100 fully distributed, airgapped, true elastic grid, all installed from SO ISO. There are no firewalls in play at the moment. We have exceeded all hardware requirements.

After a couple weeks of uptime, I am unable to join new sensors to the grid after multiple attempts, with the error given on the new forward nodes being "attempt to authenticate with salt-master failed". I have restarted the salt-master service a handful of times with symptoms being the same. I then rebooted the manager, to find that the following containers will not start (uptime > 4 hours). 

- curator
- elastalert
- elasticsearch
- filebeat
- fleet
- idstools
- kibana
- logstash
- mysql
- playbook
- redis
- soctopus

Attempting to start these with _so-service-restart_ fails with "attempt to authenticate with salt-master failed".

Troubleshooting further, it appears that a couple things could be causing this. According to the salt-master log, our 20+ searchnodes and handful of sensors are attempting to authenticate ~ every 30 seconds. Authentication is accepted by salt-master. Our dev and test grids are not seeing these seemingly excess authentication requests. 

Further, the managers salt-minion logs have the following errors:
Error while bringing up minion for multi-master. Is master at mymanager responding?
Could not pull CA from salt mine
/var/cache/salt/master/minions/mymanager_manager/mine.p does not exist

I am also unable to run any salt commands effectively, including test.ping, and receive the "salt request timed out, master not responding."

Recommendations on how to troubleshoot this issue would be highly appreciated. I suspect that the missing mine.p file could be at the core of it, but can't say for certain.

**Comment by darth-drew** [comment]
Team,

The issue continues. I have started salt-minion and salt-master in debug mode but the logs do not appear to point to whatever is preventing the salt commands from completing. I am seeing a number of exceptions in the salt-minion logs, but I believe they are caused by the message timeouts. I have powered down half of the searchnodes in order to attempt to free up the salt-master but I am still unable to get commands to succeed, unless using --async, which appears to work only some of the time. 

Due to our environment, screenshots and logs are difficult to transfer, but please let me know what information is required to help troubleshoot this issue.

**Comment by m0duspwnens** [comment]
On the manager, is the salt-master service running continuously, or is it restarting / crashing?

If you tail `/opt/so/salt/log/master` on the manager then go to a search node and run `salt-call state.apply common -l debug`, do you see the log line in `/opt/so/salt/log/master` indicating that the search node requested to authenticate and that it was accepted? On the search node, does it then tell you that it failed to authenticate with the salt-master? From the debug output running the state on the search node(towards the top where the connection is established), are there any clues as for the issue? If you run the same on the manager, does it also show the same behavior as not being able to authenticate?

Using netcat, are you able to confirm that the search node can connect to both ports 4505 and 4506 on the manager?

On your manager and search node, can you run `salt --versions` and confirm that Salt 3004 is running that the versions match for the dependency versions on the manager and search node? We are specifically interested in the versions of `PyZMQ` and `ZMQ`, but want to double check all of them.

On the manager, are there any errors in `/var/log/messages` (CentOS) or `/var/log/syslog` (Ubuntu) that might highlight the issue with the salt master service or OS in general? For example, is the salt-master service getting OOMkilled or are there any errors about too many open files?

**Comment by darth-drew** [comment]
On the manager, is the salt-master service running continuously, or is it restarting / crashing?

- Salt-master service is running continuously. Restarted 8 hours ago on this host while troubleshooting.



If you tail /opt/so/salt/log/master on the manager then go to a search node and run salt-call state.apply common -l debug, do you see the log line in /opt/so/salt/log/master indicating that the search node requested to authenticate and that it was accepted? 

- Yes, I see authentication requests from this node being accepted in masterlog. I do not see a way to determine which are automated vs manual.

On the search node, does it then tell you that it failed to authenticate with the salt-master? 

- Yes.

From the debug output running the state on the search node(towards the top where the connection is established), are there any clues as for the issue?

- None are immediately jumping out. There is a warning that Insecure logging configuration is detected. Public key is loaded and then 7 attempts with a SaltReqTimeoutError. More often though, it seems to make it further.. to decrypt the master AES key, determine pillar cache, initialize new async Auth, tries to connect 3 times then fails. Then there is a Reinit ZMQ Socket message time out before salt_unhandled exception in global handler in File: "/bin/salt-call, line 11, in <module>", load entry point ('salt==3004', 'console_scripts', 'saltcall')() and numerous python errors.

 If you run the same on the manager, does it also show the same behavior as not being able to authenticate?
- Yes

Using netcat, are you able to confirm that the search node can connect to both ports 4505 and 4506 on the manager?
- Yes

On your manager and search node, can you run salt --versions and confirm that Salt 3004 is running that the versions match for the dependency versions on the manager and search node? We are specifically interested in the versions of PyZMQ and ZMQ, but want to double check all of them.
- Confirmed

On the manager, are there any errors in /var/log/messages (CentOS) or /var/log/syslog (Ubuntu) that might highlight the issue with the salt master service or OS in general? For example, is the salt-master service getting OOMkilled or are there any errors about too many open files?
- Negative. Nothing is jumping out. 

NOTE: I have also tried a number of troubleshooting steps on the salt-master service. Most recently https://docs.saltproject.io/en/latest/topics/troubleshooting/index.html#salt-master-stops-responding 

I apologize I am not able to provide log files directly.

**Comment by m0duspwnens** [comment]
On the manager, does the md5 of `/etc/salt/pki/master/master.pub` match the md5 of `/etc/salt/pki/minion/minion_master.pub` on the search node?

If you run, from the manager, `salt-key -L`, are there any denied, unaccepted or rejected keys?

Depending on the answers for the previous questions, we can try to run the `salt-master` in [open_mode](https://docs.saltproject.io/en/latest/ref/configuration/master.html#open-mode) if you are comfortable testing that. If you make that config change to `/etc/salt/master`, then restart the `salt-master` service, do you see the same behavior from the remote nodes? Once done testing, remove the `open_mode` option from the master config and restart the `salt-master` service to disable `open_mode` again.

From the manager, if you run `salt-call pillar.get global:soversion -l debug`, do you see it attempt to connect to `127.0.0.1` or does it attempt to reach the IP of the manager? If you run the same from a remote node, is it attempting to connect to the correct IP of the manager?

**Comment by darth-drew** [comment]
Thanks for the response. /etc/salt/pki/minion/minion_master.pub does not exist on the searchnode. There are no unaccepted keys. I will try to test in open_mode tomorrow.. ran out of time today. salt-call on the manager does attempt to connect to localhost address (important to note here that we are seeing saltReqTimeouts here). salt-call on the remote node does attempt to connect to the correct IP.

**Comment by darth-drew** [comment]
Running in open_mode does not seem to impact this issue. I attempted to add a new search node to this manager but it had the same issue as before. Also, attempting to restart containers on the manager itsself still fails with timeout errors.

**Comment by m0duspwnens** [comment]
The only thing I can think of that is left for us to try is to regenerate the keys for the master and minions.

On the manager, run as root:

-  Stop the `salt-minion` service: `systemctl stop salt-minion`
-  Stop the `salt-master` service: `systemctl stop salt-master`
-  Verify that there aren't any salt processes still running and kill them if they are: `ps -ef | grep salt`
-  Create a directory somewhere to backup the current keys
-  Move the keys on the manager to the backup directory: `mv /etc/salt/pki/* <directory_created_above>`
-  Start the `salt-master` service: `systemctl start salt-master`
-  Start the `salt-minion` service: `systemctl start salt-minion`
-  Check the pending salt keys waiting to be accepted and verify if your minions are listed in here. You should only see keys listed under `Unaccepted Keys`. Run the following: `salt-key -L`
-  Accept the key for your manager: `salt-key -ya <manager's_key>`
-  Verify if the `salt-minion` running on the manager can communicate with the `salt-master`:  `salt-call pillar.get global:soversion`
-  If the previous step was successful, we can move on to fixing the other minions. If it wasn't successful, it may be worth testing the following steps on one remote node.

On the non-manager nodes:

-  Stop the `salt-minion` service: `systemctl stop salt-minion`
-  Verify that there aren't any salt processes still running and kill them if they are: `ps -ef | grep salt`
-  Remove the master's public key: `rm /etc/salt/pki/minion/minion_master.pub`
-  Start the `salt-minion` service. This takes 30 seconds to start: `systemctl start salt-minion`

Back on the manager:

-  Accept the key for the minion in the previous steps: `salt-key -ya <minion's_key>`

Back on the same node as before:

-  Start the `salt-minion` service: `systemctl start salt-minion`
-  Verify it can communicate with the manager: `salt-call pillar.get global:soversion`

**Comment by darth-drew** [comment]
Thank you for the suggestions. My notes are in sub-bullets below.

**TLDR: **Perhaps this all revolves around the logstash.nodes error?**

On the manager, run as root:

- Stop the salt-minion service: systemctl stop salt-minion
- Stop the salt-master service: systemctl stop salt-master
- Verify that there aren't any salt processes still running and kill them if they are: ps -ef | grep salt
    - There was a minion process here that I was unable to kill/pkill. rebooted and started from the top.
- Create a directory somewhere to backup the current keys
- Move the keys on the manager to the backup directory: mv /etc/salt/pki/* <directory_created_above>
- Start the salt-master service: systemctl start salt-master
- Start the salt-minion service: systemctl start salt-minion
    - status of salt minion was showing failed after multiple attempts with little additional info. rebooted and it came back up. 
- Check the pending salt keys waiting to be accepted and verify if your minions are listed in here. You should only see keys listed under Unaccepted Keys. Run the following: salt-key -L
    - The **ONLY** key here was the managers minion key. It was unaccepted.
- Accept the key for your manager: salt-key -ya <manager's_key>
- Verify if the salt-minion running on the manager can communicate with the salt-master: salt-call pillar.get global:soversion
    - This command completes successfully, however I am seeing errors when attempting to run a highstate. "Data failed to compile.. rendering SLS 'logstash.nodes' failed.. see master log for details.." 
    - Exception occurs with error: jinja variable 'dict object' has no attribute 'myreceivert01_receiver'
    - I commented out the receiver nodes in their .sls file as well as the local logstash configs, verified these salt-keys are nowhere to be found, and reran salt \*manager state.highstate. No change.. sugestions on how to remediate this problem alone?
- If the previous step was successful, we can move on to fixing the other minions. If it wasn't successful, it may be worth testing the following steps on one remote node.
    - Attempted the following set of steps with the same receiver node that logstash complained about but the receiver key never showed up in the salt-key output, and salt-minion output on the receiver node is "Error while bringing up minion for multi-master. Is master at mymanager responding?". I then checked the Salt-minion output on the manager node and it is the same.
    - Attempted to restart in open_mode and restart salt-minions. Salt-minion output is the same. the only key in salt-key -L output is the managers. Removed "open_mode: True" from /etc/salt/master and restarted master service.  **Perhaps this all revolves around the logstash.nodes error?**

On the non-manager nodes:

Stop the salt-minion service: systemctl stop salt-minion
Verify that there aren't any salt processes still running and kill them if they are: ps -ef | grep salt
Remove the master's public key: rm /etc/salt/pki/minion/minion_master.pub
Start the salt-minion service. This takes 30 seconds to start: systemctl start salt-minion
Back on the manager:

Accept the key for the minion in the previous steps: salt-key -ya <minion's_key>
Back on the same node as before:

Start the salt-minion service: systemctl start salt-minion
Verify it can communicate with the manager: salt-call pillar.get global:soversion

**Comment by m0duspwnens** [comment]
@darth-drew 
```
- Verify if the salt-minion running on the manager can communicate with the salt-master: salt-call pillar.get global:soversion
    - This command completes successfully, however I am seeing errors when attempting to run a highstate. "Data failed to compile.. rendering SLS 'logstash.nodes' failed.. see master log for details.." 
    - Exception occurs with error: jinja variable 'dict object' has no attribute 'myreceivert01_receiver'
    - I commented out the receiver nodes in their .sls file as well as the local logstash configs, verified these salt-keys are nowhere to be found, and reran salt \*manager state.highstate. No change.. sugestions on how to remediate this problem alone?
```

Since the command `salt-call pillar.get global:soversion` completed successfully, we can now say that the `salt-minion` on the manager can communicate with the `salt-master` on the manager. The two errors mentioned in bullet points 1 and 2, require information from the other nodes in the grid, and are occurring since the other nodes in the grid are unable to communicate with the manager.


```
- If the previous step was successful, we can move on to fixing the other minions. If it wasn't successful, it may be worth testing the following steps on one remote node.
    - Attempted the following set of steps with the same receiver node that logstash complained about but the receiver key never showed up in the salt-key output, and salt-minion output on the receiver node is "Error while bringing up minion for multi-master. Is master at mymanager responding?". I then checked the Salt-minion output on the manager node and it is the same.
    - Attempted to restart in open_mode and restart salt-minions. Salt-minion output is the same. the only key in salt-key -L output is the managers. Removed "open_mode: True" from /etc/salt/master and restarted master service.  **Perhaps this all revolves around the logstash.nodes error?**
```

Regarding the first bullet point and the error that was seen on the receiver node: `"Error while bringing up minion for multi-master. Is master at mymanager responding?"`

This would tend to indicate that the node is unable to communicate with the `salt-master` running on the manager. This could occur for a few reason, but not limited to:

- The `salt-master` service is not running on the manager
- The firewall on the manager is blocking the connection or there is a firewall on the network blocking it
- The receiver's minion config, located at `/etc/salt/minion`, has the incorrect manager hostname, the hostname set in there is resolving to an incorrect IP, or there is some other network related issue preventing the Salt traffic from making it from the receiver to the manager

Since it is mentioned that the only key listed in the output from `salt-key -L` is the manager's key, it reaffirms that none of the other nodes are able to communicate with the `salt-master` running on the manager. We would expect to see all the minions keys listed after a short period of time. The `logstash.nodes` error is not causing this issue, but instead is being seen because this issue.

Regarding ` I then checked the Salt-minion output on the manager node and it is the same.`
If I am understanding this correctly, in a previous step we confirmed that the `salt-minion` on the manager is able to communicate with the `salt-master` on the manager, but this indicates that it has stopped. Is that the case, or are you still able to successfully run `salt-call pillar.get global:soversion` on the manager?

**Reply by darth-drew** [reply]
Thanks for breaking this down for me Josh. So what I am seeing is `salt-call pillar.get global:soversion` on the manager completing successfully, however the minion log indicates "Minion unable to successfully connect to a Salt Master" every 5 minutes almost exactly. Salt-master service has been running for 3 days. The fact that I am seeing this in the salt-minion logs on the manager to me indicates there is something other then your three bullets above, which is where I am stuck. I am seeing the salt-minion service restarting every 5m as well.

**Comment by m0duspwnens** [comment]
From the manager, does running `salt-call pillar.get global:soversion` complete successfully every time, even if you see the `salt-minion` log on the manager indicate "Minion unable to successfully connect to a Salt Master"? I would expect that you are seeing also seeing a message such as "Error while bringing up minion for multi-master. Is master at <my_salt_master> responding?" approximately every 30 seconds. The error "Minion unable to successfully connect to a Salt Master" likely should only be seen once following a restart of the service and "Error while bringing up minion for multi-master. Is master at <my_salt_master> responding?" will likely be seen every 30 seconds as the `salt-minion` is attempting to authenticate with the `salt-master` running on the manager.

If you check the timestamp on the manager for the file `/opt/so/log/salt/lasthighstate` what is the date? This will tell us the last time the manager was able to run a highstate.

The reason the `salt-minion` service is restarting every 5 minutes is due to a cronjob that is run as root. If you run `crontab -e` as root, you can remove the two lines below to prevent it from auto restarting as it may help with troubleshooting since we won't have to worry about the minion restarting while testing. Once everything is working properly again, the cronjob will be added back from a highstate.
```
# SALT_CRON_IDENTIFIER:so-salt-minion-check
*/5 * * * * /usr/sbin/so-salt-minion-check -q
```
The script the cronjob runs is used to determine if the `salt-minion` is successfully able to apply states (and thus communicate with the `salt-master`) by checking the last time the file `/opt/so/log/salt/state-apply-test` was touched. This file should be touched every 5 minutes by a scheduled job that runs the state `salt.minion-state-apply-test` where the sole purpose of this state is to touch that file. If the state is unable to run and touch the file, it is either due to another state run (like a highstate) being hung, or the `salt-minion` can't communicate with the `salt-master` on the manager. By default, if the file `/opt/so/log/salt/state-apply-test` hasn't been touched within 3600 seconds, then the script will restart the `salt-minion` service.

Perhaps it is possible that the `salt-minion` is giving up waiting for the `salt-master` to accept the minion's key, by default this is 5 seconds. This is normally fine, but if the manager / salt-master is under high load, then we may need to adjust this. On the manager, we can edit `/etc/salt/minion` and add `auth_timeout: 30` to change it to 30 seconds. While we are in here, we can change the log file log level to debug to give us a little more information. In `/etc/salt/minion` change `log_level_logfile: error` to `log_level_logfile: debug`. After that, we can run `systemctl stop salt-minion` then check to see if any `salt-minion` or `salt-call` processes are still running and kill them if so. Once that is done, we can restart the `salt-minion` service with `systemctl start salt-minion`. After that, tail `/opt/so/log/salt/minion` and verify if the minion is able to connect to the `salt-master`. If the connection is still failing, we should see the message in the log `[DEBUG   ][18411] SaltReqTimeoutError, retrying. (1/7)` written every 30 seconds now instead of every 5. If the minion is able to successfully authenticate and run a highstate, then `log_level_logfile: debug` will be changed back to `log_level_logfile: error` and the `salt-minion` service will restart since that is a config setting that we manage to reduce log size.

Is it possible that there are some other processes or controls on the manager that would be interfering with the `salt-minion` to `salt-master` communication?

**Comment by darth-drew** [comment]
All on the manager, `salt-call pillar.get global:soversion` does seem to complete successfully every time. I am no longer seeing the "Error while bringing up minion for multi-master. Is master at <my_salt_master> responding?" entry in the minion log. At this point, I am uncertain what change may have caused this. 

The file /opt/so/log/salt/lasthighstate was last touched on March 1st.

so-salt-minion-check has been commented out of the crontab.

The file /opt/so/log/salt/state-apply-test has not been touched since March 3rd.

After making the adjustments to /etc/salt/minion for auth_timeout and log_level, and restarting the service, It seems like the connection is working with logstash error present. Lasthighstate log is still not touched.

As far as other processes that would be interfering.. nothing that I am aware of. We have commented out the last 4 lines of /etc/ssh/sshd_config, but that change was made on Feb10.
